src/test.rs

```diff
@@ -0,0 +1,117 @@
+//! Test driver
+use crate::utils::ParseError;
+use std::str::FromStr;
+extern crate structopt;
+use structopt::StructOpt;
+mod basic;
+mod pfm;
+mod testutils;
+
+/// Test Struct
+pub struct Test {
+    pub name: String,
+    pub description: String,
+    pub call: fn() -> bool,
+    pub subtests: Vec<Test>,
+    pub is_subtest: bool,
+}
+
+/// TestResult
+pub enum TestResult {
```

Comment:
Similarly, you could define your own generic result type here, whose error would be the error that has been propagated up. This will provide a way for us to print a message to the user for explaining why the test failed. The reason I don't think you need `Skipped` is because the user will have specified which tests have been skipped, and propagating that information up isn't necessary.

Reply:
I disagree here -- I think `Skipped` is fairly important to send to the CLI. It improves the readability of test output when there is a large number of tests and many test harnesses provide a `Skipped` message.

Reply:
My whole argument is based around making things clear to the user, so I think you are right.

Reply:
Yep, I think Skipped is good, the real `perf test` has Skipped messages like that so I was just trying to make it as similar as possible. Also, if I was using this, I would love if there was some kind of `--json` option that wrote the output of the tests to a JSON file. Maybe that's a little silly, but I think it would be really easy to do, so maybe I'll do it...

Reply:
That doesn't sound silly at all. I think it's a great idea.